libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#ifdef ENABLE_SCRIPTING
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <RC_Channel/RC_Channel.h>
+#include "AP_MotorsMatrix.h"
+
+class AP_MotorsMatrix_6DoF_Scripting : public AP_MotorsMatrix {
+public:
+
+    /// Constructor
+    AP_MotorsMatrix_6DoF_Scripting(uint16_t loop_rate, uint16_t speed_hz = AP_MOTORS_SPEED_DEFAULT) :
+        AP_MotorsMatrix(loop_rate, speed_hz)
+    {
+        if (_singleton != nullptr) {
+            AP_HAL::panic("AP_MotorsMatrix 6DoF must be singleton");
+        }
+        _singleton = this;
+    };
+
+    // get singleton instance
+    static AP_MotorsMatrix_6DoF_Scripting *get_singleton() {
+        return _singleton;
+    }
+
+    // output_to_motors - sends minimum values out to the motors
+    void output_to_motors() override;
+
+    void set_roll_pitch(float roll_deg, float pitch_deg) override;
+
+    // add_motor using raw roll, pitch, throttle and yaw factors, to be called from scripting
+    void add_motor(int8_t motor_num, float roll_factor, float pitch_factor, float yaw_factor, float throttle_factor, float forward_factor, float right_factor, bool is_3D, uint8_t testing_order);
+
+    // if the expected number of motors have been setup then set as initalized
+    bool init(uint8_t expected_num_motors) override;
+
+protected:
+    // output - sends commands to the motors
+    void output_armed_stabilizing() override;
+
+    // nothing to do for setup, scripting will mark as initalized when done
+    void setup_motors(motor_frame_class frame_class, motor_frame_type frame_type) override {};
+
+    float _throttle_factor[AP_MOTORS_MAX_NUM_MOTORS];     // each motors contribution to up thrust
+    float _forward_factor[AP_MOTORS_MAX_NUM_MOTORS];      // each motors contribution to forward thrust
+    float _right_factor[AP_MOTORS_MAX_NUM_MOTORS];        // each motors contribution to right thrust
+
+    // true if motor is 3D, it can go from -Spin max to +Spin max, if false motor is can go from Spin min to Spin max
+    bool  _motor_3D[AP_MOTORS_MAX_NUM_MOTORS];
```

Comment:
it's just naming but I guess this is a motor that uses a "3D" propeller so it can provide thrust when reversed?  Could we use a name that is more specific in terms of what it's unusual ability is.  something like "_can_reverse_thrust[]" or "_reversible[]" or something like this.  Sorry, I don't want to start you off on a guessing game of what I think is an acceptable name but when I see, "_motor_3D" I think that perhaps this is a 3D position of a motor or a 3D model of a motor.  It doesn't scream out to me that it is a motor that can provide reverse thrust.